### ISSUE TITLE
issue with kafka bitnami helm chart not creating bitnami/kafka/config folder

### ISSUE THREAD

**Issue by faisal-basha** [bitnami/charts#4336]
**Which chart**:
Helm Chart Kafka 

**Describe the bug**
While installing the helm chart the kafka broker pods are not being created as they are crashing. The logs say that
mkdir -p /bitnami/kafka/config permission denied.

**Expected behavior**
kafka pod creation along with zookeeper.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
v3.4.0+g7090a89

```

- Output of `kubectl version`:

```
Client Version: v1.19.3
Server Version: v1.19.3

```

**Additional context**
It is not at all allowing me to create the pods as the code in the statefulsets.yaml template tries to create the folder within the docker image which is not allowed.
<img width="749" alt="kafka-bug" src="https://user-images.githubusercontent.com/68839822/98871057-b9bb9880-248d-11eb-9587-3e14ba6e8e41.png">

**Comment by javsalgar** [comment]
Hi,

Thank you for using Bitnami. This looks like an issue with the persistent volume provider. Could you provide more details about the cluster you are using and the storage provider? We test the chart with major Kubernetes distributions and storage providers and didn't find issues.

**Comment by faisal-basha** [comment]
Hi, 

This screenshot should sum it up all. Please see the attached image. I am using a persistent volume which uses storage class called Standard and it is mounted with ISILON DMC where kafka would store all of its logs and everything.

<img width="960" alt="kafka-bugnew" src="https://user-images.githubusercontent.com/68839822/99148024-e09ce900-269e-11eb-91db-a02690266477.png">

**Comment by alemorcuq** [comment]
Hi, @faisal-basha.

Have you tried deploying the chart with `volumePermissions.enabled` set to `true`? It deploys an init container that changes the permissions of the mounted volume before launching the main container.

You can see more about this here: https://github.com/bitnami/charts/tree/master/bitnami/kafka#volume-permissions-parameters

**Comment by faisal-basha** [comment]
I have enabled volumePermissions in the config file and also I have tried sending it as a parameter while installing the helm chart. I see the same error message. even after that I get the permissions issue is it something to do with CIFS mount? I have a CIFS mounted with my persistent volume.

**Comment by alemorcuq** [comment]
It is probably something to do with your CIFS volume. Maybe you are missing some `mountOption` to make it work?

**Comment by faisal-basha** [comment]
Hi there, I have created a test pod with hostPath as my /mnt/data which is my cifs mount and /data is the volume within the container to which the container will write data. Now when I exec into the pod and create a file within the /data directory I am able to see that the same file has been created in the /mnt/data directory. So I know that I dont have an issue with my cifs mounts.

**Comment by javsalgar** [comment]
Hi,

In your test you said that you were able to create a file. My question is, were you using a non-root container for the test? Some volumes have issues with non-root containers.

**Comment by faisal-basha** [comment]
Hi,

I got through that issue by adding a statement to create the folder within that init-permissions container. But I have another issue with logs folder which wont start my kafka brokers. It is stuck at this now:

<img width="960" alt="bitnamikafka" src="https://user-images.githubusercontent.com/68839822/100009856-d9858180-2de8-11eb-8d91-0adf9edf866e.png">

**Comment by javsalgar** [comment]
Hi,

It definitely looks to me that you have some sort of podsecuritypolicy in place that is causing some issues with the permissions. Could you provide more details about that? Do you have ReadOnlyFilesystem enabled? That could explain the issue.

**Comment by faisal-basha** [comment]
Hi please take a look at my configuration  which is the values.yaml file. I have removed all the comments, also I have not given my extraDeploy which is a deployment template by itself which has private database information which I dont want to expose. Please take a look at this config values yaml file which has settings for the podSecuritylevel.


image:
  registry: docker.io
  repository: bitnami/kafka
  tag: 2.6.0-debian-10-r88
  pullPolicy: IfNotPresent
  pullSecrets: []
  debug: false

clusterDomain: cluster.local
commonLabels: {}
commonAnnotations: {}

heapOpts: -Xmx1024m -Xms1024m

deleteTopicEnable: false
autoCreateTopicsEnable: true
logFlushIntervalMessages: 10000
logFlushIntervalMs: 1000
logRetentionBytes: _1073741824
logRetentionCheckIntervalMs: 300000
logRetentionHours: 168
logSegmentBytes: _1073741824
logsDirs: /opt/mydam_data/kafka/logs
maxMessageBytes: _1000012
defaultReplicationFactor: 3
offsetsTopicReplicationFactor: 3
transactionStateLogReplicationFactor: 3
transactionStateLogMinIsr: 3
numIoThreads: 8
numNetworkThreads: 3
numPartitions: 3
numRecoveryThreadsPerDataDir: 1
socketReceiveBufferBytes: 102400
socketRequestMaxBytes: _104857600
socketSendBufferBytes: 102400
zookeeperConnectionTimeoutMs: 6000

command:
  - /scripts/setup.sh
args:

extraEnvVars: []
extraVolumes: []
extraVolumeMounts: []

auth:
  clientProtocol: plaintext
  interBrokerProtocol: plaintext
  enabled: false
  saslMechanisms: plain,scram-sha-256,scram-sha-512
  saslInterBrokerMechanism: plain
  tlsEndpointIdentificationAlgorithm: https
  jaas:
    interBrokerUser: admin
    interBrokerPassword: ""
listeners: []
advertisedListeners: []
allowPlaintextListener: true
interBrokerListenerName: INTERNAL
replicaCount: 1
updateStrategy: RollingUpdate
podLabels: {}
podAnnotations: {}
priorityClassName: ""
affinity: {}
nodeSelector: {}
tolerations: []
podSecurityContext:
  fsGroup: 1001
  runAsUser: 1001

containerSecurityContext: {}

resources:
  limits: {}
  requests: {}

livenessProbe:
  enabled: true
  initialDelaySeconds: 10
  timeoutSeconds: 5
  # failureThreshold: 3
  # periodSeconds: 10
  # successThreshold: 1
readinessProbe:
  enabled: true
  initialDelaySeconds: 5
  failureThreshold: 6
  timeoutSeconds: 5
  # periodSeconds: 10
  # successThreshold: 1

customLivenessProbe: {}
customReadinessProbe: {}

pdb:
  create: true
  maxUnavailable: 1

sidecars: {}

service:
  type: ClusterIP
  port: 9092
  internalPort: 9093
  externalPort: 9094
  nodePorts:
    client: ""
    external: ""
  loadBalancerSourceRanges: []
  annotations: {}

externalAccess:
  enabled: true
  autoDiscovery:
    enabled: false
    image:
      registry: docker.io
      repository: bitnami/kubectl
      tag: 1.17.13-debian-10-r21
      pullPolicy: IfNotPresent
      pullSecrets: []
    resources:
      limits: {}
      requests: {}

  service:
    type: NodePort
    loadBalancerSourceRanges: []
    nodePorts: [30011] 
    #30012,30013]
    domain: 10.126.0.92 (fake ip address)
    annotations: {}

persistence:
  enabled: true
  storageClass: "standard"
  accessModes:
    - ReadWriteOnce
  size: 100Gi

logPersistence:
  enabled: true
  storageClassName: "standard"
  storageClass: "standard"
  accessModes:
    - ReadWriteOnce
  size: 100Gi
  annotations: {}

volumePermissions:
  enabled: true
  image:
    registry: docker.io
    repository: bitnami/minideb
    tag: buster
    pullPolicy: Always
    pullSecrets: []
  resources:
    limits: {}
    requests: {}

serviceAccount:
  create: true
rbac:
  create: true

metrics:
  kafka:
    enabled: false
    image:
      registry: docker.io
      repository: bitnami/kafka-exporter
      tag: 1.2.0-debian-10-r277
      pullPolicy: IfNotPresent
      pullSecrets: []
    extraFlags: {}
    resources:
      limits: {}
      requests: {}
    service:
      type: ClusterIP
      port: 9308
      nodePort: ""
      loadBalancerSourceRanges: []
      annotations:
        prometheus.io/scrape: "true"
        prometheus.io/port: "{{ .Values.metrics.kafka.service.port }}"
        prometheus.io/path: "/metrics"
  jmx:
    enabled: false
    image:
      registry: docker.io
      repository: bitnami/jmx-exporter
      tag: 0.14.0-debian-10-r64
      pullPolicy: IfNotPresent
      pullSecrets: []

    resources:
      limits: {}
      requests: {}

    service:
      type: ClusterIP
      port: 5556
      nodePort: ""
      loadBalancerSourceRanges: []
      annotations:
        prometheus.io/scrape: "true"
        prometheus.io/port: "{{ .Values.metrics.jmx.service.port }}"
        prometheus.io/path: "/"

    whitelistObjectNames:
      - kafka.controller:*
      - kafka.server:*
      - java.lang:*
      - kafka.network:*
      - kafka.log:*

    config: |-
      jmxUrl: service:jmx:rmi:///jndi/rmi://127.0.0.1:5555/jmxrmi
      lowercaseOutputName: true
      lowercaseOutputLabelNames: true
      ssl: false
      {{- if .Values.metrics.jmx.whitelistObjectNames }}
      whitelistObjectNames: ["{{ join "\",\"" .Values.metrics.jmx.whitelistObjectNames }}"]
      {{- end }}

  serviceMonitor:
    enabled: false

zookeeper:
  enabled: true
  metrics:
    enabled: true
  auth:
    enabled: false
  persistence:
    enabled: false
externalZookeeper:
  servers: []

**Comment by faisal-basha** [comment]
Both, fsGroup and RunAsUser are set to 1001 within the yaml file and they read via the stateful.yaml file within the chart where it creates folders within the containers.

**Comment by faisal-basha** [comment]
Hi,

Sorry I forgot to mention that I was getting an error after the helm install command failed. I had made changes to the statefulsets.yaml file i.e updated it with the latest code for some reason my previous releases had old version of your chart.  Now that I updated the statefulset.yaml file with the code for log Persistence, and I tried to install it again. It says  Error: statefulsets.apps "kafka-release" already exists where Kafka-release is the name of my helm chart. I think this is what is avoiding the pods to be recreated how can I overcome this issue form helm installation ?

**Comment by faisal-basha** [comment]
HI I got a new version of this chart I mean got the updated version of this version and ran it, but it gives the same issue with the above configuration. Not sure whats causing this issue.

**Comment by faisal-basha** [comment]
I figured why this issue was happening. This happens because of enabling logPersistence, there is a bug in your code when enabling log Persistence this error happens. Now when I set enabled: false for logPersistence in the values.yaml that error goes away for my kafka broker. There is some kind of bug for you guys to fix..

**Comment by faisal-basha** [comment]
I think this where you error is :

{{- if not .Values.logPersistence.enabled }}
        - name: logs
          emptyDir: {}

if the logsPersistence is not enabled only then it will create the logs emptyDir I think you should have it regardless true or false.

**Comment by javsalgar** [comment]
Hi,

Thanks for the input. I think this confirms what I was thinking: your cluster must have a ReadOnlyFilesystem policy enabled cluster wide, which makes all the pods have this policy. That would explain the fact that only after adding this emptyDir volume the pod runs without issues. We test our chart will all major Kubernetes distributions and this issue does not appear. However, our charts are not yet fully compatible with ReadOnlyFilesystems. It's good to know that in the case of rabbitmq it works with this value enabled. Could you confirm with your cluster administrator that you have a policy like this set cluster wide?
